File: server/src/packet/PacketType.cpp
#include "PacketType.h"

namespace packet {

    uint8_t packet::PacketType::toUInt8(header header) {
        return static_cast<uint8_t>(header);
    }

    packet::PacketType::header packet::PacketType::toPacketType(uint8_t header) {
        static std::unordered_map<uint8_t, PacketType::header> packetTypeMap = {
                {static_cast<uint8_t>(PacketType::header::OPEN),            PacketType::header::OPEN},
                {static_cast<uint8_t>(PacketType::header::CLOSE),           PacketType::header::CLOSE},
                {static_cast<uint8_t>(PacketType::header::ACK_OK),          PacketType::header::ACK_OK},
                {static_cast<uint8_t>(PacketType::header::ACK_ERR),         PacketType::header::ACK_ERR},
                {static_cast<uint8_t>(PacketType::header::ALIVE),           PacketType::header::ALIVE},

                {static_cast<uint8_t>(PacketType::header::OPEN_ENCR),       PacketType::header::OPEN_ENCR},
                {static_cast<uint8_t>(PacketType::header::PUBLIC_KEY),      PacketType::header::PUBLIC_KEY},
                {static_cast<uint8_t>(PacketType::header::SYMMETRIC_KEY),   PacketType::header::SYMMETRIC_KEY},
                {static_cast<uint8_t>(PacketType::header::TEST_KEY),        PacketType::header::TEST_KEY},

                {static_cast<uint8_t>(PacketType::header::OPEN_PROT),       PacketType::header::OPEN_PROT},
                {static_cast<uint8_t>(PacketType::header::SET_NAME),        PacketType::header::SET_NAME},
                {static_cast<uint8_t>(PacketType::header::ADD_FOLLOWER),    PacketType::header::ADD_FOLLOWER},
                {static_cast<uint8_t>(PacketType::header::NEW_FOLLOWED),    PacketType::header::NEW_FOLLOWED},
                {static_cast<uint8_t>(PacketType::header::REMOVE_FOLLOWER), PacketType::header::REMOVE_FOLLOWER},
                {static_cast<uint8_t>(PacketType::header::REMOVE_FOLLOWED), PacketType::header::REMOVE_FOLLOWED},
                {static_cast<uint8_t>(PacketType::header::MY_LOCATION),     PacketType::header::MY_LOCATION},
                {static_cast<uint8_t>(PacketType::header::LOCATION),        PacketType::header::LOCATION},
        };

        return packetTypeMap.count(header) ? packetTypeMap.at(header) : PacketType::header::UNDEFINED;
    }

    std::string PacketType::toString(PacketType::header header) {
        static std::unordered_map<PacketType::header, std::string> packetTypeMap = {
                {PacketType::header::OPEN,            "OPEN"},
                {PacketType::header::CLOSE,           "CLOSE"},
                {PacketType::header::ACK_OK,          "ACK_OK"},
                {PacketType::header::ACK_ERR,         "ACK_ERR"},
                {PacketType::header::ALIVE,           "ALIVE"},

                {PacketType::header::OPEN_ENCR,       "OPEN_ENCR"},
                {PacketType::header::PUBLIC_KEY,      "PUBLIC_KEY"},
                {PacketType::header::SYMMETRIC_KEY,   "SYMMETRIC_KEY"},
                {PacketType::header::TEST_KEY,        "TEST_KEY"},

                {PacketType::header::OPEN_PROT,       "OPEN_PROT"},
                {PacketType::header::SET_NAME,        "SET_NAME"},
                {PacketType::header::ADD_FOLLOWER,    "ADD_FOLLOWER"},
                {PacketType::header::NEW_FOLLOWED,    "NEW_FOLLOWED"},
                {PacketType::header::REMOVE_FOLLOWER, "REMOVE_FOLLOWER"},
                {PacketType::header::REMOVE_FOLLOWED, "REMOVE_FOLLOWED"},
                {PacketType::header::MY_LOCATION,     "MY_LOCATION"},
                {PacketType::header::LOCATION,        "LOCATION"},
        };

        return packetTypeMap.count(header) ? packetTypeMap.at(header) : "UNDEFINED";
    }
}

File: client/src/main/java/com/github/koraxiss/PacketType.java
package main.java.com.github.koraxiss;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public enum PacketType {
    _OPEN,
    _CLOSE,

    _OPEN_ENCR,
    _PUBLIC_KEY,
    _SYMMETRIC_KEY,
    _TEST_KEY,

    _OPEN_PROT,
    _SET_NAME,
    _ADD_FOLLOWER,
    _NEW_FOLLOWED,
    _REMOVE_FOLLOWER,
    _REMOVE_FOLLOWED,
    _MY_LOCATION,
    _LOCATION,
    _ACK_OK,
    _ACK_ERR,
    _ALIVE,
    _OTHER;

    public static Map<Integer, PacketType> packetTypeMap = new HashMap<>();
    public static Map<PacketType, Integer> reversedMap = new HashMap<>();

    static {
        Map<Integer, PacketType> map = new HashMap<>();
        map.put(0, PacketType._OPEN);
        map.put(1, PacketType._CLOSE);

        map.put(8, PacketType._OPEN_ENCR);
        map.put(9, PacketType._PUBLIC_KEY);
        map.put(10, PacketType._SYMMETRIC_KEY);
        map.put(11, PacketType._TEST_KEY);

        map.put(16, PacketType._OPEN_PROT);
        map.put(17, PacketType._SET_NAME);
        map.put(18, PacketType._ADD_FOLLOWER);
        map.put(19, PacketType._NEW_FOLLOWED);
        map.put(20, PacketType._REMOVE_FOLLOWER);
        map.put(21, PacketType._REMOVE_FOLLOWED);
        map.put(22, PacketType._MY_LOCATION);
        map.put(23, PacketType._LOCATION);
        map.put(2, PacketType._ACK_OK);
        map.put(3, PacketType._ACK_ERR);
        map.put(4, PacketType._ALIVE);

        packetTypeMap = Collections.unmodifiableMap(map);

        for(Integer i: packetTypeMap.keySet())
            reversedMap.put(packetTypeMap.get(i), i);
    }
}
